File: CS217 Data Structures and Algorithm Analysis (H)/std/Lab13/B/Accepted1.java
import java.io.*;
import java.util.*;

/* Ioana, 1 BFS*/

public class Accepted1
{
	static class AdjList
	{
		int m_count;
		int[] m_members;
		public AdjList() {
			m_count = 0;
		}
	}

	static int N, M;
	static int[] levels;
	static int[] queue;
	static int[][] edges;
	static AdjList[] neighbours; 
	
	static int longestshortest(int startvertex)
	{
		for (int i = 0; i<N;++i)
			levels[i] = 0;
		
		int first = 0, last = 0;  
		queue[0] = startvertex;
		levels[startvertex] = 1;  
    	while (first <= last) {
    		int crt = queue[first];
    		for (int i = 0; i < neighbours[crt].m_count; ++i) {
    			int crtneighb = neighbours[crt].m_members[i];
    			if (levels[crtneighb] == 0) {
    				queue[++last] = crtneighb;
    				levels[crtneighb] = levels[crt]+1;
    			}

    		}   
    		++first;               
    	}
    	if (last < N-1)
    		return 0;
    	return levels[queue[last]];
	}

	static void readAndProcess() {
	    try {
	        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    
            StringTokenizer stringtokenizer = new StringTokenizer(br.readLine());
            N = Integer.parseInt(stringtokenizer.nextToken());
            M = Integer.parseInt(stringtokenizer.nextToken());
		
		    levels = new int[N];
	        queue = new int[N];
	        edges = new int[M][2];
		    neighbours = new AdjList[N];
		
		    for (int i = 0; i < N; ++i)
			    neighbours[i] = new AdjList();
		
		    for (int i = 0; i < M; ++i){
		        stringtokenizer = new StringTokenizer(br.readLine());
	            int r1 = Integer.parseInt(stringtokenizer.nextToken());
                int r2 = Integer.parseInt(stringtokenizer.nextToken());
	            r1--;
	            r2--;
	            edges[i][0]=r1;
	            edges[i][1]=r2;
	            neighbours[r1].m_count++;
	             neighbours[r2].m_count++;
	        }
	    } catch(Exception exception) { }
        
	
		for (int i = 0; i < N; ++i) { 
	        neighbours[i].m_members = new int[neighbours[i].m_count];
	        neighbours[i].m_count = 0;
	    }
	    
	    for (int i = 0; i < M; ++i) {
	        neighbours[edges[i][0]].m_members[neighbours[edges[i][0]].m_count++] = edges[i][1];
	        neighbours[edges[i][1]].m_members[neighbours[edges[i][1]].m_count++] = edges[i][0];
	    }
	}
	
	public static void main(String[] args) throws FileNotFoundException
	{
		readAndProcess();
		int estimatediam = longestshortest(0)-1;
    
		if (estimatediam == -1) 
			System.out.println("-1\n");
		else {
			double log2 = Math.log(estimatediam) / Math.log(2);
			System.out.println((int)Math.ceil(log2)+1); 
		}
	}
}

File: CS217 Data Structures and Algorithm Analysis (H)/std/Lab13/B/accepted1.py
#! /usr/bin/python3

import sys
import math
    
n,m = map(int,input().split())

ledges = [None]*m
counts = [0]*n
neighb = [None]*n
levels = [None]*n
queue = [None]*n

def longestshortest(startvertex):
    for i in range(0, n):
        levels[i] = 0
    first = last = 0  
    queue[0] = startvertex
    levels[startvertex] = 1
    while (first <= last):
        #print(first)
        crt = queue[first]
        for i in range(0,counts[crt]):
            crtneighb = neighb[crt][i]
            if (levels[crtneighb] == 0):
                last = last + 1
                queue[last] = crtneighb
                levels[crtneighb] = levels[crt]+1
        first = first + 1
    if (last < n-1):
        return 0;
    return levels[queue[last]]    
        
ctr = 0

for i in range (0,m):
    i,j = map(int,input().split())
    i = i-1
    j = j-1
    
    ledges[ctr] = (i,j)
    counts[i] = counts[i] + 1
    counts[j] = counts[j] + 1
    ctr = ctr + 1
    
for i in range(0,n):
    neighb[i] = [None] * counts[i]
    counts[i] = 0
    

for [i,j] in ledges:
    neighb[i][counts[i]]=j
    counts[i] = counts[i] + 1
    neighb[j][counts[j]]=i
    counts[j] = counts[j] + 1
    
    
estimatediam = longestshortest(0)

if (estimatediam == 0): 
    print('-1');
else:
    print(math.ceil(math.log2(estimatediam-1))+1) 
  
     

